<?xml version="1.0"?>
<launch>
  <arg name="uri1" default="radio://0/80/2M/E7E7E7E701" />
  <arg name="uri2" default="radio://0/80/2M/E7E7E7E702" />
  <arg name="uri3" default="radio://0/80/2M/E7E7E7E703" />
  <arg name="frame1" default="crazyflie1" />
  <arg name="frame2" default="crazyflie2" />
  <arg name="frame3" default="crazyflie3" />
  <arg name="ip" default="172.16.1.1" />
  <arg name="port" default="3883" />
  <include file="$(find crazyflie_driver)/launch/crazyflie_server.launch"></include>

  <group ns="$(arg frame1)">
    <node pkg="crazyflie_driver" type="crazyflie_add" name="crazyflie_add" output="screen">
      <param name="uri" value="$(arg uri1)" />
      <param name="tf_prefix" value="$(arg frame1)" />
      <param name="enable_logging" value="False" /> <!--Logging is disabled to reduce unnecessary bandwith consumption-->
      <param name="enable_logging_imu" value="False" />
      <param name="enable_logging_temperature" value="False" />
      <param name="enable_logging_magnetic_field" value="False" />
      <param name="enable_logging_pressure" value="False" />
      <param name="enable_logging_battery" value="True" />
      <param name="enable_logging_packets" value="True" />
      <rosparam>
        genericLogTopics: ["log1"]
        genericLogTopicFrequencies: [100]
        genericLogTopic_log1_Variables: ["stateEstimate.x", "stateEstimate.y", "stateEstimate.z"]
      </rosparam>
    </node>
    <node name="pose" pkg="crazyflie_demo" type="publish_external_position_vrpn.py" output="screen"> 
      <param name="topic" value="/$(arg frame1)/vrpn_client_node/$(arg frame1)/pose" />
    </node>
    <node pkg="vrpn_client_ros" type="vrpn_client_node" name="vrpn_client_node" output="screen">
      <rosparam subst_value="true">
        server: $(arg ip)
        port: $(arg port)
        update_frequency: 100.0
        frame_id: /world
        child_frame_id: $(arg frame1)
        use_server_time: false
        broadcast_tf: true
        refresh_tracker_frequency: 1.0
      </rosparam>
    </node>
  </group>
  <group ns="$(arg frame2)">
    <node pkg="crazyflie_driver" type="crazyflie_add" name="crazyflie_add" output="screen">
      <param name="uri" value="$(arg uri2)" />
      <param name="tf_prefix" value="$(arg frame2)" />
      <param name="enable_logging" value="False" /> <!--Logging is disabled to reduce unnecessary bandwith consumption-->
      <param name="enable_logging_imu" value="False" />
      <param name="enable_logging_temperature" value="False" />
      <param name="enable_logging_magnetic_field" value="False" />
      <param name="enable_logging_pressure" value="False" />
      <param name="enable_logging_battery" value="True" />
      <param name="enable_logging_packets" value="True" />
      <rosparam>
        genericLogTopics: ["log1"]
        genericLogTopicFrequencies: [100]
        genericLogTopic_log1_Variables: ["stateEstimate.x", "stateEstimate.y", "stateEstimate.z"]
      </rosparam>
    </node>
    <node name="pose" pkg="crazyflie_demo" type="publish_external_position_vrpn.py" output="screen"> 
      <param name="topic" value="/$(arg frame2)/vrpn_client_node/$(arg frame2)/pose" />
    </node>
    <node pkg="vrpn_client_ros" type="vrpn_client_node" name="vrpn_client_node" output="screen">
      <rosparam subst_value="true">
        server: $(arg ip)
        port: $(arg port)
        update_frequency: 100.0
        frame_id: /world
        child_frame_id: $(arg frame2)
        use_server_time: false
        broadcast_tf: true
        refresh_tracker_frequency: 1.0
      </rosparam>
    </node>
  </group>
  <group ns="$(arg frame3)">
    <node pkg="crazyflie_driver" type="crazyflie_add" name="crazyflie_add" output="screen">
      <param name="uri" value="$(arg uri3)" />
      <param name="tf_prefix" value="$(arg frame3)" />
      <param name="enable_logging" value="False" /> <!--Logging is disabled to reduce unnecessary bandwith consumption-->
      <param name="enable_logging_imu" value="False" />
      <param name="enable_logging_temperature" value="False" />
      <param name="enable_logging_magnetic_field" value="False" />
      <param name="enable_logging_pressure" value="False" />
      <param name="enable_logging_battery" value="True" />
      <param name="enable_logging_packets" value="True" />
      <rosparam>
        genericLogTopics: ["log1"]
        genericLogTopicFrequencies: [100]
        genericLogTopic_log1_Variables: ["stateEstimate.x", "stateEstimate.y", "stateEstimate.z"]
      </rosparam>
    </node>
    <node name="pose" pkg="crazyflie_demo" type="publish_external_position_vrpn.py" output="screen"> 
      <param name="topic" value="/$(arg frame3)/vrpn_client_node/$(arg frame3)/pose" />
    </node>
    <node pkg="vrpn_client_ros" type="vrpn_client_node" name="vrpn_client_node" output="screen">
      <rosparam subst_value="true">
        server: $(arg ip)
        port: $(arg port)
        update_frequency: 100.0
        frame_id: /world
        child_frame_id: $(arg frame3)
        use_server_time: false
        broadcast_tf: true
        refresh_tracker_frequency: 1.0
      </rosparam>
    </node>
  </group>
</launch>
